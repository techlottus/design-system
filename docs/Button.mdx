---
name: Button
menu: Components
---

import { Playground, Props } from 'docz'
import Button from '../packages/Button'

# Button

Button component docs ðŸ“„

<Props of={Button} />

## Basic Usage

<Playground>
  <Button>Primary btn</Button>
  <Button variant='secondary'>Secondary btn</Button>
  <Button variant='terciary'>Terciary btn</Button>
  <Button variant='disabled'>Disabled btn</Button>
</Playground>