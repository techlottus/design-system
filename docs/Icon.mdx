---
name: Icon
menu: Components
---

import { Props } from 'docz'
import Icon from '../packages/Icon';
import Nav from '../packages/Nav';


# Icon
Icon component docs

## Props
<Props of={Icon}/>

## Basic
```jsx
<Icon size={32} icon="icon-twitter" />
<Icon size={32} icon="icon-facebook" />
<Icon size={32} icon="icon-linkedin" />
<Icon size={32} icon="icon-instagram" />
```
<Nav>
  <Nav.Item>
    <Icon size={32} icon="icon-twitter" />
  </Nav.Item>
  <Nav.Item>
    <Icon size={32} icon="icon-facebook" />
  </Nav.Item>
  <Nav.Item>
    <Icon size={32} icon="icon-linkedin" />
  </Nav.Item>
  <Nav.Item>
    <Icon size={32} icon="icon-instagram-glyph" />
  </Nav.Item>
</Nav>

## With gap
```jsx
<Icon size={32} gap={2} icon="icon-arrows" />
<Icon size={32} gap={4} icon="icon-flag" />
<Icon size={32} gap={6} icon="icon-money" />
<Icon size={32} gap={8} icon="icon-cart" />
```
<Nav>
  <Nav.Item>
    <Icon size={32} gap={2} icon="icon-arrows" />
  </Nav.Item>
  <Nav.Item>
    <Icon size={32} gap={4} icon="icon-flag" />
  </Nav.Item>
  <Nav.Item>
    <Icon size={32} gap={6} icon="icon-money" />
  </Nav.Item>
  <Nav.Item>
    <Icon size={32} gap={8} icon="icon-cart" />
  </Nav.Item>
</Nav>