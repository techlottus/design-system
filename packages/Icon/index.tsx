import * as React from 'react';
import { IProps } from '../common/props';
import classNames from 'classnames';
import style from './icon.module.pcss';

export interface IIconProps extends IProps {
  size?: number;
  gap?: number;
  icon?: string;
}

const Icon: React.SFC<IIconProps> = ({ size, gap, className, icon }) => {
  const classes = classNames(className, 'betomic', icon);
  return (<div className={style.iconWrapper} style={{
    width: size,
    height: size,
    padding: gap,
  }}>
    <i style={{
      fontSize: (size - gap),
    }} className={classes}></i>
  </div>);
};

Icon.defaultProps = {
  size: 16,
  gap: 0,
};

export default Icon;
