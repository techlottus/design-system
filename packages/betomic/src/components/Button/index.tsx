
import React, { ButtonHTMLAttributes, MouseEvent } from "react";
import cn from "classnames";

const Sizes = {
  sm: cn("px-4 py-1.5 text-sm h-8"),
  md: cn("px-5 py-3 text-base h-12"),
  lg: cn("px-6 py-4 text-xl h-15")
};

type Size = keyof typeof Sizes;

const SizesIcon = {
  sm: cn("p-1.5 w-8 h-8"),
  md: cn("p-3.5 w-12 h-12"),
  lg: cn("p-4 w-15 h-15")
};

const commonClasses = cn("rounded", "flex", "justify-center", "items-center", "font-montserrat", "font-bold", "focus:outline-none", "focus:border-primary-500", "focus:ring-2", "focus:ring-primary-200");

const Variants = {
  primary: cn("bg-primary-500", "text-white", "border", "border-transparent", "hover:bg-primary-600", "focus:bg-primary-600", "focus:border", commonClasses),
  secondary: cn("bg-surface-600", "text-white", "border", "border-transparent", "hover:bg-surface-900", "focus:bg-surface-900", "focus:border", commonClasses),
  tertiary: cn("bg-white", "text-surface-600", "border-2", "border-surface-500", "hover:bg-surface-500", "hover:text-white", "focus:bg-surface-500", "focus:text-white",
    "dark:bg-transparent", "dark:text-white", "dark:border-white", "dark:hover:bg-surface-50", "dark:hover:text-surface-600", "dark:focus:bg-surface-50", "dark:focus:text-surface-600", "dark:focus:border-primary-500", commonClasses)
};

type Variant = keyof typeof Variants;

interface IButton extends ButtonHTMLAttributes<HTMLButtonElement> {
  block?: boolean;
  children: React.ReactNode;
  className?: string;
  /**
   * If set to `true`, the button will display a centered loading spinner instead of its contents, and the button will be disabled.
   * The width of the button is not affected by the value of this prop.
   *
   * @default false
   */
  loading?: boolean;
  disabled?: boolean;
  icon?: boolean;
  size?: Size;
  type?: "button" | "submit" | "reset";
  variant?: Variant;
  style?: React.CSSProperties;
  onClick?: (event: MouseEvent<HTMLElement>) => void;
};

const Button: React.FC<IButton> = (props: IButton) => {
  const { 
    block = false,
    children,
    className,
    disabled = false,
    loading = false,
    icon = false,
    size = "md",
    type,
    variant = "primary",
    onClick,
    ...restProps 
  } = props;
  const isDisabled = disabled || loading;

  const buttonDisabledClasses = cn("bg-surface-50", "border", "border-transparent", "text-surface-300", "cursor-not-allowed", "hover:bg-surface-50", "focus:bg-surface-50", commonClasses);

  const renderChildren = () => {
    const sizes = {
      sm: 'h-4 w-4',
      md: 'h-5 w-5',
      lg: 'h-7 w-7'
    };
    return [
      loading && <svg className={cn("animate-spin text-surface-400", sizes[size])} xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24">
        <circle className="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" stroke-width="4"></circle>
        <path className="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"></path>
      </svg>,
      !loading && children,
    ];
  }

  const buttonClasses = cn(
    className, {
      [Sizes[size]]: !icon,
      [SizesIcon[size]]: icon,
      [Variants[variant]]: !isDisabled,
      [buttonDisabledClasses]: isDisabled,
      ["w-full"]: block,
    }
  );

  return (
    <button
      type={type}
      className={buttonClasses}
      disabled={disabled}
      onClick={onClick}
      {...restProps}
    >
      {renderChildren()}
    </button>
  );
}

export default Button;