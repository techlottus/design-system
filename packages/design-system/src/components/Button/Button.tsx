import cn from "classnames";
import { buttonTypes } from "../Types";

const sizes = {
  xsm: cn("px-4 py-3 text-sm rounded"),
  sm: cn("py-4 px-8 text-base rounded-lg"),
  md: cn("py-5 px-9 text-xl rounded-lg"),
  lg: cn("py-6 px-12 text-xl rounded-lg"),
};
type Size = keyof typeof sizes;

const variants = {
  primary: cn(
    "bg-neutral-900 hover:bg-neutral-800 focus:bg-neutral-600 active:bg-neutral-600",
    "text-neutral-100",
    "border border-neutral-900 active:border-0 active:ring-0 focus:border-neutral-900 focus:ring-2 focus:ring-neutral-900 focus:outline-none"
  ),
  outlined: cn(
    "bg-transparent hover:bg-neutral-200 active:bg-neutral-400 focus:bg-neutral-100",
    "text-neutral-900",
    " border border-transparent focus:border-neutral-900 focus:ring-2 focus:ring-neutral-900 focus:outline-none"
  ),
  text: cn(
    "bg-transparent hover:bg-neutral-200 active:bg-neutral-400 ",
    "text-neutral-900 hover:text-neutral-900",
    "active:ring-0 active:border-0 focus:border focus:border-neutral-900 focus:ring-1 focus:ring-neutral-900 "
  ),
  "outlined-negative": cn(
    "bg-transparent text-neutral-100 border border-neutral-100  ",
    "hover:bg-neutral-100 hover:border-neutral-900 hover:text-neutral-900",
    "active:ring-0 active:border-0 active:bg-neutral-400 active:text-neutral-100 ",
    "focus:border focus:border-neutral-100 focus:ring-1 focus:ring-neutral-100 "
  ),
};
type Variant = keyof typeof variants;

const commonClasses = cn(
  "flex",
  "relative",
  "min-w-40 max-w-full",
  "font-bold font-principal",
  "items-center justify-center"
);
const isDisabled = {
  primary: cn("cursor-not-allowed", "bg-neutral-400", "text-neutral-100"),
  text: cn("cursor-not-allowed text-neutral-400"),
  outlined: cn(
    "cursor-not-allowed",
    "bg-transparent",
    "text-neutral-400",
    "border border-neutral-400"
  ),
  "outlined-negative": cn(
    "cursor-not-allowed",
    "bg-transparent",
    "text-neutral-200",
    "border border-neutral-200"
  ),
};

interface IButton extends buttonTypes {
  size?: Size;
  variant?: Variant;
}

const Button: React.FC<IButton> = (props: IButton) => {
  const {
    id = "btn",
    className = "",
    disabled = false,
    children,
    size = "md",
    variant = "primary",
    onClick,
    type = "button",
    ...restProps
  } = props;
  const buttonClasses = cn(commonClasses, className, {
    [sizes[size]]: true,
    [variants[variant]]: !disabled,
    [isDisabled[variant]]: disabled,
    ["cursor-pointer"]: !disabled,
  });
  return (
    <button
      id={id}
      type={type}
      className={buttonClasses}
      onClick={onClick}
      {...restProps}
    >
      {children}
    </button>
  );
};

export default Button;
