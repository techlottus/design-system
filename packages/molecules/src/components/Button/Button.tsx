import cn from "classnames";
import { ButtonSize, ButtonType, ButtonVariant } from "../Types/Button.types";
import React from "react";
const sizes: ButtonSize = {
  xsm: cn("px-4 py-3 text-sm rounded"),
  sm: cn("py-4 px-8 text-base rounded-lg"),
  md: cn("py-5 px-9 text-xl rounded-lg"),
  lg: cn("py-6 px-12 text-xl rounded-lg"),
};

const variants: ButtonVariant = {
  primary: cn(
    "bg-primary-500 hover:bg-primary-600 focus:bg-primary-700 active:bg-primary-700",
    "text-surface-100",
    "border-2 border-primary-500  hover:border-primary-600 active:border-primary-700 active:ring-0 focus:border-primary-500 focus:ring-2 focus:ring-primary-50 focus:outline-none"
  ),
  outlined: cn(
    "bg-transparent hover:bg-surface-200 active:bg-surface-400 focus:bg-surface-100",
    "text-surface-900",
    " border border-surface-900 focus:border-surface-900 focus:ring-2 focus:ring-surface-900 focus:outline-none"
  ),
  text: cn(
    "bg-transparent hover:bg-surface-200 active:bg-surface-400 ",
    "text-surface-900 hover:text-surface-900",
    "active:ring-0 active:border-0 focus:border focus:border-surface-900 focus:ring-1 focus:ring-surface-900 "
  ),
  "outlined-negative": cn(
    "bg-transparent text-surface-100 border border-surface-100  ",
    "hover:bg-surface-100 hover:border-surface-900 hover:text-surface-900",
    "active:ring-0 active:border-0 active:bg-surface-400 active:text-surface-100 ",
    "focus:border focus:border-surface-100 focus:ring-1 focus:ring-surface-100 "
  ),
};

const commonClasses: string = cn(
  "flex",
  "relative",
  "min-w-40 max-w-full",
  "font-bold font-headings",
  "items-center justify-center"
);

const disabledVariants: ButtonVariant = {
  primary: cn("cursor-not-allowed", "bg-surface-200", "text-surface-400"),
  text: cn("cursor-not-allowed text-surface-400"),
  outlined: cn(
    "cursor-not-allowed",
    "bg-transparent",
    "text-surface-400",
    "border border-surface-400"
  ),
  "outlined-negative": cn(
    "cursor-not-allowed",
    "bg-transparent",
    "text-surface-200",
    "border border-surface-200"
  ),
};

const Button: React.FC<ButtonType> = (props: ButtonType) => {
  const {
    className = "",
    disabled = false,
    size = "md",
    variant = "primary",
    children,
    onClick,
  } = props;

  const buttonClasses: string = cn(commonClasses, className, [sizes[size]], {
    [variants[variant]]: !disabled,
    [disabledVariants[variant]]: disabled,
    ["cursor-pointer"]: !disabled,
  });

  return (
    <button className={buttonClasses} onClick={onClick}>
      {children}
    </button>
  );
};

export default Button;
