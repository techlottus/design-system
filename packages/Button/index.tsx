import * as React from 'react';
import Icon from '../Icon';

// Button Props
const sizes = {
  xs: '',
  sm: '',
  md: 'px-8 py-3 text-base',
  lg: 'px-10 py-5 text-xl',
};

const buttonVariant = {
  primary: `
    bg-primary-400
    hover:bg-primary-700
    inline-flex
    items-center
    rounded
    text-white
    font-bold
    font-montserrat
  `,
  secondary: `
    bg-surface-500
    hover:bg-surface-800
    inline-flex
    items-center
    rounded
    text-white
    font-bold
    font-montserrat
  `,
  terciary: `
    bg-transparent
    hover:bg-surface-400
    hover:text-white
    border-2
    border-surface-400
    rounded
    inline-flex
    items-center
    font-bold
    font-montserrat
  `,
  disabled: `
    bg-surface-100
    rounded
    text-surface-300
    font-bold
    font-montserrat
    cursor-not-allowed
  `,
};

interface ButtonProps {
  size: 'xs'|'sm'|'md'|'lg';
  variant: 'primary'|'secondary'|'terciary'|'disabled';
  children: any;
  className?: string;
}

const Button: React.SFC<ButtonProps> = ({ size, variant, children, className }: ButtonProps) => (
  <button className={`${className} ${sizes[size]} ${buttonVariant[variant]}`}>{children}</button>
);

Button.defaultProps = {
  size: 'md',
  variant: 'primary',
  className: '',
};

export default Button;
