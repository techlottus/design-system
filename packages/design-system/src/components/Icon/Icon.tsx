import cn from "classnames";
import { IconType } from "../Types/Icon.types";

const Icon: React.FC<IconType> = (props: IconType) => {
  const { iconName, className = "", variant = "solid" } = props;
  return (
    <span
      className={cn(className, {
        ["font-icons-outlined"]: variant === "outlined",
        ["font-icons-solid"]: variant === "solid",
      })}
    >
      <i>{iconName}</i>
    </span>
  );
};

export default Icon;
