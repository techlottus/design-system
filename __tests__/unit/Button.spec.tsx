import * as React from 'react';
import { shallow } from 'enzyme';
import Button from '../../packages/Button/index';
import Icon from '../../packages/Icon/index';

describe('Button.tsx', () => {
  it('should match the snapshot', () => {
    const vm = shallow(<Button variant='primary' size='md'>Button</Button>);
    expect(vm).toMatchSnapshot();
  });

  describe('Properties', () => {
    it('the component Button has to be a "button" tag', () => {
      const vm = shallow(<Button variant='primary' size='md'>Button</Button>);
      expect(vm.find('button')).toHaveLength(1);
    });

    it('the prop children has to be a string', () => {
      const vm = shallow(<Button variant='primary' size='md'>Button</Button>);
      expect(typeof vm.props().children).toBe('string');
    });

    it('the prop children accepts an Icon component as children', () => {
      const vm = shallow(
        <Button variant='secondary' size='md'>
          Button <Icon icon='fab fa-amazon'/>
        </Button>);
      expect(typeof vm.props().children).toBe('object');
    });

    it('test click event', () => {
      const mockCallBack = jest.fn();
      const vm = shallow(
      <Button variant='secondary' size='md' onClick={mockCallBack}>
        Click!
      </Button>);
      vm.find('button').simulate('click');
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });
  });
});
