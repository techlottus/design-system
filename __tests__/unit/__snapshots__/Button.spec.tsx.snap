// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.tsx should match the snapshot 1`] = `
<button
  className=" px-8 py-3 text-base 
    bg-primary-400
    hover:bg-primary-700
    rounded
    focus:outline-none
    focus:shadow-primary-200
    focus:bg-primary-700
    inline-flex
    items-center
    text-white
    font-bold
    font-montserrat
  "
>
  Button
</button>
`;
